returnn/search.py

```diff
@@ -14,6 +14,156 @@
+            returnn_config, log_verbosity, device
+        )
+
+        self.rqmt = {
+            "gpu": 1 if device == "gpu" else 0,
+            "cpu": cpu_rqmt,
+            "mem": mem_rqmt,
+            "time": time_rqmt,
+        }
+
+    def tasks(self):
+        yield Task("create_files", mini_task=True)
+        yield Task("run", rqmt=self.rqmt)
+
+    def create_files(self):
+        config = self.returnn_config
+
+        config.config["load"] = self.model_checkpoint.ckpt_path
+        config.write(self.out_returnn_config_file.get_path())
+
+        cmd = [
+            tk.uncached_path(self.returnn_python_exe),
+            os.path.join(tk.uncached_path(self.returnn_root), "rnn.py"),
+            self.out_returnn_config_file.get_path(),
+        ]
+
+        util.create_executable("rnn.sh", cmd)
+
+        # check here if model actually exists
+        assert os.path.exists(
+            tk.uncached_path(self.model_checkpoint.index_path)
+        ), "Provided model does not exists: %s" % str(self.model_checkpoint)
+
+    def run(self):
+        call = [
+            tk.uncached_path(self.returnn_python_exe),
+            os.path.join(tk.uncached_path(self.returnn_root), "rnn.py"),
+            self.out_returnn_config_file.get_path(),
+        ]
+        sp.check_call(call)
+
+    def create_returnn_config(self, returnn_config, log_verbosity, device):
+        """
+        Creates search RETURNN config
+        :param ReturnnConfig returnn_config: object representing RETURNN config
+        :param int log_verbosity: RETURNN log verbosity
+        :param str device: RETURNN device, cpu or gpu
+        :rtype: ReturnnConfig
+        """
+        assert device in ["gpu", "cpu"]
+        original_config = returnn_config.config
+        assert "network" in original_config
+
+        config = {
+            "search_output_file_format": "py",
+            "search_output_file": self.out_search_file,
+            "need_data": False,
+            "search_do_eval": 0,
+        }
+
+        config.update(copy.deepcopy(original_config))  # update with the original config
+
+        # override always
+        config["task"] = "search"
+        config["max_seq_length"] = 0
+
+        if "search_data" in original_config:
+            config["search_data"] = {
+                **original_config["search_data"].copy(),
+                **self.search_data,
+            }
+        else:
+            config["search_data"] = self.search_data
+
+        post_config = {
+            "device": device,
+            "log": ["./returnn.log"],
+            "log_verbosity": log_verbosity,
+        }
+
+        post_config.update(copy.deepcopy(returnn_config.post_config))
+
+        res = copy.deepcopy(returnn_config)
+        res.config = config
+        res.post_config = post_config
+
+        return res
+
+    @classmethod
+    def hash(cls, kwargs):
+        d = {
+            "returnn_config": kwargs["returnn_config"].hash(),
```

Comment:
We should fix the hashing behavior of ReturnnConfig completely so that calling this explicitly is not needed anymore, but this can be done in another PR as this affects other jobs as well.